src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1438,26 +1459,31 @@ mongoc_uri_parse (mongoc_uri_t *uri, const char *str, bson_error_t *error)
    BSON_ASSERT (str);
 
    if (*str) {
+      // Check for valid end of hostname delimeter (skip slash if necessary)
+      if (*str != '/' && *str != '?') {
+         MONGOC_URI_ERROR (error, "%s", "Expected end of hostname delimiter");
+         goto error;
+      }
+
       if (*str == '/') {
+         // Try to parse database.
          str++;
          if (*str) {
             if (!mongoc_uri_parse_database (uri, str, &str)) {
                MONGOC_URI_ERROR (error, "%s", "Invalid database name in URI");
                goto error;
             }
          }
+      }
 
-         if (*str == '?') {
-            str++;
-            if (*str) {
-               if (!mongoc_uri_parse_options (uri, str, false /* from DNS */, error)) {
-                  goto error;
-               }
+      if (*str == '?') {
```

Comment:
```suggestion
      else if (*str == '?') {
```

Recommend using `else if` to denote related and exclusive conditional branches.

Reply:
I think keeping this conditional as an `if` is necessary to ensure that a URI can first be parsed for its database name and then separately be checked for options (e.g. `mongodb://localhost:27017/test?replicaset=foo`). Changing to an `else if` will parse the database name but then skip parsing the options completely. Let me know if there's a misunderstanding on my end.

Reply:
Good catch. I neglected to consider that `str` at this point depends on the preceeding `str++` and `mongoc_uri_parse_database()` even in the `*str == '/' -> true` case.